### PR TITLE
fix: visit a space just after delete

### LINKED ISSUE #352: Clicking on Visit Space from Home after deletion does not work
To reproduce:
- From home, delete a space (you have only one space)
- Click on Visit A Space (second button)
- Visit A Space (maybe the same)

Expected Result: Visit the Space in preview mode

Result: The application returns to Home.

### PR DESCRIPTION
**Pull request by pyphilia**
close #352 

I added more details in the issue itself. In this PR I got rid of the `deleted` property and used `onSuccess` callback when deleting a space to redirect the page.

### CODE REVIEW

File: src/components/space/SpaceScreen.js [hunk 1]
```diff
@@ -82,7 +80,7 @@ class SpaceScreen extends Component {
     }).isRequired,
     history: PropTypes.shape({
       length: PropTypes.number.isRequired,
-      replace: PropTypes.func.isRequired,
+      goBack: PropTypes.func.isRequired,
```
Comment (juancarlosfarah) [inline src/components/space/SpaceScreen.js:83]: Do you need `goBack`?

File: src/components/space/SpaceHeader.js [hunk 2]
```diff
@@ -93,10 +97,16 @@ class SpaceHeader extends Component {
...
-    const { space } = this.props;
+    const {
+      space,
+      history: { goBack },
+    } = this.props;
     const { saved, id } = space;
     if (saved) {
-      return <DeleteButton spaceId={id} />;
+      const onSuccess = () => {
+        goBack();
```
Comment (juancarlosfarah) [inline src/components/space/SpaceHeader.js:107]: Are you sure `goBack` works for all cases?

File: src/actions/space.js [hunk 3]
```diff
@@ -246,14 +244,14 @@ const saveSpace = async ({ space }) => async dispatch => {
...
-const clearSpace = () => dispatch => {
+const clearSpace = () => (dispatch) => {
   dispatch(clearPhase());
   return dispatch({
     type: CLEAR_SPACE,
   });
 };
 
-const deleteSpace = ({ id }) => dispatch => {
+const deleteSpace = ({ id }, onSuccess) => (dispatch) => {
```
Comment (juancarlosfarah) [inline src/actions/space.js:254]: Why don't you put `onSuccess` inside the object?

File: src/actions/space.js [hunk 4]
```diff
@@ -276,6 +274,9 @@ const deleteSpace = ({ id }) => dispatch => {
         i18n.t(ERROR_DELETING_MESSAGE)
       );
     } else {
+      // eslint-disable-next-line no-unused-expressions
+      onSuccess?.();
```
Comment (juancarlosfarah) [inline src/actions/space.js:278]: Crazy syntax. :-P

**Review by juancarlosfarah — CHANGES_REQUESTED** [review]
Thanks for the PR! What was wrong with the delete? Just some minor comments.

File: src/components/space/SpaceHeader.js (hunk 2 above)
Comment (pyphilia) [reply src/components/space/SpaceHeader.js:107]: This only includes cases where the delete button in SpaceHeader is used. I cannot know where we will use it of course, but I can assume that the SpaceHeader is only used when viewing a space at `space/id`. Thus the goBack function will redirect to whatever page we come from (there are only two cases: Home page and Saved Spaces). Let me know if we should be more specific than `goBack`

File: src/actions/space.js (hunk 3 above)
Comment (pyphilia) [reply src/actions/space.js:254]: I prefer to keep the object and the callback apart. That way the payload for the back-end won't include onSuccess as well (though it doesn't change much).

**Comment by pyphilia** [comment]
The problem is the following: Even though the space was correctly removed from the local database, the state was keeping the space in `space.current.content` with an additional property `deleted: true`. This property was used to redirect the user to the saved spaces if the space was deleted from `SpaceScreen`. 
Therefore, when a space was fetched for visiting, the space already existed in the state as deleted, so the redirection was triggered and lead us to the home screen.

I answered to your questions, let me know if it now clear.

File: src/components/space/SpaceHeader.js (hunk 2 above)
Comment (juancarlosfarah) [reply src/components/space/SpaceHeader.js:107]: @pyphilia, we can leave it as is for now.

File: src/actions/space.js (hunk 3 above)
Comment (juancarlosfarah) [reply src/actions/space.js:254]: OK.

**Review by juancarlosfarah — APPROVED** [review]

**Comment by juancarlosfarah** [comment]
This PR is good for @pyphilia, but I'm seeing a weird error on CodeShip. Something to do with the linter.

**Pull request merged**